src/geometry/detail/VGNavCollection.hh

```diff
@@ -0,0 +1,204 @@
+struct VGNavCollection<Ownership::reference, MemSpace::host>
+{
+    using NavState = vecgeom::cxx::NavigationState;
+
+    NavState* ptr = nullptr;
+
+    // Obtain reference from host memory
+    void operator=(VGNavCollection<Ownership::value, MemSpace::host>& other);
+    // Get the navigation state for a given thread
+    NavState& at(int, ThreadId id) const;
+    //! True if the collection is assigned/valiid
+    explicit operator bool() const { return static_cast<bool>(ptr); }
+};
+#endif
+
+//---------------------------------------------------------------------------//
+// DEVICE MEMSPACE
+//---------------------------------------------------------------------------//
+/*!
+ * Delete a VecGeom pool.
+ *
+ * Due to VecGeom macros, the definition of this function can only be compiled
+ * from a .cc file.
+ */
+struct NavStatePoolDeleter
+{
+    using arg_type = vecgeom::cxx::NavStatePool*;
+    void operator()(arg_type) const;
+};
+
+#ifndef __CUDA_ARCH__
+//---------------------------------------------------------------------------//
+/*!
+ * Manage a pool of device-side geometry states.
+ *
+ * Construction and destruction of the NavStatePool has to be in a host
+ * compilation unit due to VecGeom macro magic. We hide this class to keep
+ * NavStatePool and smart pointer usage from the NVCC device compiler.
+ */
+template<>
+struct VGNavCollection<Ownership::value, MemSpace::device>
+{
+    using UPNavStatePool
+        = std::unique_ptr<vecgeom::cxx::NavStatePool, NavStatePoolDeleter>;
+
+    UPNavStatePool pool;
+    void*          ptr  = nullptr;
+    size_type      size = 0;
+
+    // Resize based on geometry params and state size
+    void resize(int max_depth, size_type size);
+    //! True if the collection is assigned/valid
+    explicit CELER_FUNCTION operator bool() const { return ptr; }
+};
+#endif
+
+//---------------------------------------------------------------------------//
+/*!
+ * Reference on-device memory owned by VGNavCollection<value, device>.
+ *
+ * The NavStatePool underpinning the storage returns a void pointer that must
+ * be manually manipulated to get a single state pointer.
+ */
+template<>
+struct VGNavCollection<Ownership::reference, MemSpace::device>
+{
+    using NavState = vecgeom::cuda::NavigationState;
+
+    void* ptr = nullptr;
+    size_type size = 0;
+
+    // Assign from device value
+    void operator=(VGNavCollection<Ownership::value, MemSpace::device>& other);
+    // Get the navigation state for the given thread
+    inline CELER_FUNCTION NavState& at(int max_depth, ThreadId thread) const;
+    //! True if the collection is assigned/valiid
+    explicit CELER_FUNCTION operator bool() const { return ptr; }
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Get the navigation state at the given thread.
+ */
+CELER_FUNCTION auto VGNavCollection<Ownership::reference, MemSpace::device>::at(
+    int max_depth, ThreadId thread) const -> NavState&
```

Comment:
Having the max_depth being 'customizable' is a source of potential problem.  The depth used here *must* be the same as the depth used during allocation.

Reply:
Yeah, true -- originally I had `max_depth` as part of the state too but it's redundant with the params. It wouldn't be "costly" to duplicate that for the VG nav collection as member data (two more ints in const memory, no big deal) but it's definitely less elegant. I can add it as state data or just add a comment about making sure it's consistent -- your call.

Reply:
it just that it is a potential source of mistake since the arguement "max_depth" *must* be the one expected or the result is wrong (I.e. it is a property of the collection not a parameter of the at call).

Reply:
You're right Philippe -- to prevent mistakes I've added `max_depth` to the state data. Partly out of laziness (changing the interface) and partly out of potential optimization (max_depth being loaded once from 'params' rather than from the state in two separate places) I'm still passing in max_depth which is assumed to come from the Params class, but a new assertion will make sure they're consistent.